ros/wavemap_ros/include/wavemap_ros/impl/wavemap_server_inl.h

```diff
@@ -13,91 +14,47 @@ namespace wavemap {
+  while (!changed_blocks.empty()) {
+    // Prepare the blocks to publish in the current iteration
+    std::unordered_set<Index3D, Index3DHash> blocks_to_publish;
+    int block_cnt = 0;
+    for (const auto& block_idx : changed_blocks) {
+      hashed_map->getBlock(block_idx).threshold();
+      blocks_to_publish.insert(block_idx);
+      if (config_.max_num_blocks_per_msg <= ++block_cnt) {
+        break;
+      }
     }
-  }
 
-  // Select the blocks to publish in the current cycle
-  // NOTE: We take the config_.max_num_blocks_per_msg most recently modified
-  //       blocks.
-  std::unordered_set<Index3D, Index3DHash> blocks_to_publish;
-  for (const auto& [_, block_idx] : changed_blocks_sorted) {
-    hashed_map->getBlock(block_idx).threshold();
-    blocks_to_publish.insert(block_idx);
-    if (static_cast<size_t>(config_.max_num_blocks_per_msg) <=
-        blocks_to_publish.size()) {
-      break;
+    // Serialize and publish the selected blocks
+    wavemap_msgs::Map map_msg;
+    map_msg.header.frame_id = config_.world_frame;
+    map_msg.header.stamp = ros::Time::now();
+    convert::mapToRosMsg(*hashed_map,
+                         map_msg.hashed_wavelet_octree.emplace_back(),
+                         blocks_to_publish, thread_pool_);
+    {
+      ZoneScopedN("publishMapRosMsg");
+      map_pub_.publish(map_msg);
     }
-  }
-
-  // If there are no blocks to publish, we're done
-  if (blocks_to_publish.empty()) {
-    return;
-  }
-
-  // Serialize and publish the selected blocks
-  wavemap_msgs::Map map_msg;
-  map_msg.header.frame_id = config_.world_frame;
-  map_msg.header.stamp = ros::Time::now();
-  convert::mapToRosMsg(*hashed_map,
-                       map_msg.hashed_wavelet_octree.emplace_back(),
-                       blocks_to_publish);
-  {
-    ZoneScopedN("publishMapRosMsg");
-    map_pub_.publish(map_msg);
-  }
 
-  // Remove the published blocks from the publication queue
-  std::for_each(
-      blocks_to_publish.begin(), blocks_to_publish.end(),
-      [&queue = block_publishing_queue_](const auto& published_block) {
-        queue.erase(published_block);
-      });
-
-  // Handle publishing of the remaining blocks
-  if (!block_publishing_queue_.empty()) {
-    ROS_WARN_STREAM("Could not publish all blocks at once. Published "
-                    << blocks_to_publish.size() << " out of "
-                    << config_.max_num_blocks_per_msg
-                    << ". Remaining in queue: "
-                    << block_publishing_queue_.size());
-
-    // If the map is being published at a fixed rate, wait for the next cycle
-    if (0.f < config_.publication_period) {
-      return;
-    }
-
-    // Otherwise finish publishing it now
-    if (republish_whole_map && config_.publication_period < 0.f) {
-      do {
-        publishHashedMap(hashed_map, false);
-      } while (!block_publishing_queue_.empty());
-    }
+    // Remove the published blocks from the publication queue
```

Comment:
Couldn't erasing use the from to variant of erase given you pick the N first entries from the set?

Reply:
Good idea, that's much cleaner. Done :)